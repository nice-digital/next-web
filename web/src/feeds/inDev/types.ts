import {
	AreaOfInterest,
	ProductGroup,
	ProductTypeAcronym,
} from "@/feeds/publications/types";

export enum FeedPath {
	AllProjects = "/gidprojects/all",
	InConsultationProjects = "/inconsultationprojects",
	ProjectDetail = "/project/",
}

export enum TopicSelectionReason {
	"Monitor" = "Monitor - awaiting publication of further literature or evidence",
	"Anticipate" = "Anticipate the topic will be of importance to patients, carers, professionals, commissioners and the health of the public to ensure clinical benefit is realised, inequalities in use addressed, and help them make the best use of NHS resources",
	"NotEligible" = "Not eligible for Health Technology Evaluation guidance",
	"FurtherDiscussion" = "Further discussion required at the Topic Selection Oversight Panel (TSOP)",
}

export enum ProjectStatus {
	"Complete" = "Complete",
	"Discontinued" = "Discontinued",
	"Suspended" = "Suspended",
	"Referred" = "Referred",
	"InProgress" = "InProgress",
	"Proposed" = "Proposed",
	"TopicSelection" = "TopicSelection",
	"ImpactedByCOVID19" = "ImpactedByCOVID19",
}

/** Not a definitive list - this may change over time */
export enum ProjectProcess {
	"MTA" = "MTA",
	"IP" = "IP",
	"TAG" = "TAG",
	"MT" = "MT",
	"DT" = "DT",
	"STAPre2018" = "STA pre-2018",
	"STA2018" = "STA 2018",
	"QSD" = "QSD",
	"CG" = "CG",
	"APG" = "APG",
	"SG" = "SG",
	"HST" = "HST",
	"MIB" = "MIB",
	"STAReview" = "STA Review",
	"PHG" = "PHG",
	"MTAReview" = "MTA Review",
	"CDFReview" = "CDF Review",
	"NG" = "NG",
	"APA" = "APA",
	"ES" = "ES",
	"SC" = "SC",
	"COVID19RapidGuideline" = "COVID-19 rapid guideline",
}

export enum ProjectGroup {
	Guideline = ProductGroup.Guideline,
	Guidance = ProductGroup.Guidance,
	Advice = ProductGroup.Advice,
	Standard = ProductGroup.Standard,
	Other = ProductGroup.Other,
}

export type GIDReference = `GID-${Uppercase<string>}${number}`;

export interface AllProjects {
	links: {
		// In theory InDev supports pagination, but in reality the gidprojects/all feed always returns all projects and a page size of 2048
		first: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		last: [
			{
				// This looks like a bug in InDev as Page=0
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
		self: [
			{
				href: "/gidprojects/All?Page=0&PageSize=2048";
			}
		];
	};
	embedded: {
		niceIndevIndevelopmentProject: Project[];
	};
}

export interface Project {
	links: {
		self: [
			{
				href: `/project/${GIDReference}`;
			}
		];
	};
	/** e.g. 01000000-0000-001B-0000-000000005402 */
	eTag: string;
	/** Project reference e.g. GID-TAG377 */
	reference: GIDReference;
	/** Acronym (usually 2 or 3 letters) for the type of the project e.g "TA", "IPG" etc */
	projectType: ProductTypeAcronym;
	/** The name of the project type e.g. "Technology appraisal guidance" for TA or "Interventional procedures guidance" for "IPG" */
	productTypeName: string;
	process: ProjectProcess | string;
	/** Project title e.g. "Atrial fibrillation - ximelagatran [ID376]" or "Clinically isolated syndrome - beta interferons and glatiramer acetate [ID109]" */
	title: string;
	status: ProjectStatus;
	projectGroup: ProjectGroup;
	/**
	 * `null` for unpublished projects or an ISO formatted date string like `2020-12-15T00:00:00`
	 */
	publishedDate: string | null;
	/**
	 * An ISO formatted date string like `2020-09-22T13:51:48.8447645`
	 */
	lastModifiedDate: string;
	/**
	 * Null for unpublished projects or an ISO formatted date string like `2020-10-05T12:27:21.5437767`
	 */
	firstGoLiveDate: string | null;
	/** The date the project was created in in dev */
	createdDate: string;
	developedAs: string | null;
	/** Almost always the same as `DevelopedAs` */
	relevantTo: string | null;
	idNumber: `${number}` | null;
	/** E.g. "BMJ Evidence Centre, BMJ Group", "ScHARR" or "Warwick Evidence" etc */
	evidenceAssessmentGroup: string | null;
	/** TODO Always seems to be an empty array */
	areasOfInterestList: [];
}

export interface InConsultationProjects {
	links: {
		self: [
			{
				href: "/inconsultationprojects?Page=0&PageSize=2048";
			}
		];
	};
	// Sometimes there are no consultations running, in which case `_embedded` is not in the feed at all
	embedded?: {
		niceIndevInconsultationProduct: Consultation[];
	};
	eTag: null;
	// Page and size is hard coded here
	page: 0;
	pageSize: 2048;
	totalItems: number;
}

export interface Consultation {
	_links: {
		self: [
			{
				href: `/guidance/${GIDReference}/consultation/html-content${
					| number
					| ""}`;
			}
		];
	};
	eTag: null;
	reference: GIDReference;
	title: string;
	consultationName: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	startDate: string;
	/** ISO formatted date like `2021-06-18T00:00:00` */
	endDate: string;
	consultationType: string;
	resourceTitleId: string;
	projectType: ProjectProcess | string;
	productTypeName: string;
	showExpressionOfInterestSubmissionQuestion: boolean;
	developedAs: null;
	relevantTo: null;
	consultationId: number;
	/** E.g. QSD, IP etc */
	process: string;
	hasDocumentsWhichAllowConsultationComments: boolean;
	isCHTE: boolean;
	/** E.g. HSCTeam or CHTETeam */
	allowedRole: string;
	firstConvertedDocumentId: null;
	firstChapterSlugOfFirstConvertedDocument: null;
	partiallyUpdatedProjectReference: null;
	origProjectReference: null;
	areasOfInterestList: [];
}

export type Link = {
	href: string;
};

type ETag = string | null;

export type IndevFile = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	id: string;
	mimeType: string;
	fileName: string;
	length: number;
	hash: string;
	name: string;
	reference: string;
	resourceTitleId: string;
	consultationDocumentId: number;
};

export type IndevConvertedDocument = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	reference: string;
	title: string;
	resourceTitleId: string;
	projectType: string;
	productTypeName: string;
};

export type IndevResource = {
	links: {
		self: [Link];
	};
	embedded?: {
		niceIndevFile?: IndevFile;
		niceIndevConvertedDocument?: IndevConvertedDocument;
	};
	eTag: ETag;
	title: string;
	level: number;
	publishedDate: string;
	externalUrl: string | null;
	showInDocList: boolean;
	textOnly: boolean;
	consultationId: number;
	consultationDocumentId: number;
	isCurrentViewableConsultationMarkup: boolean;
	convertedDocument: boolean;
	supportsComments: boolean;
	supportsQuestions?: boolean;
};

export type IndevFileResource = IndevResource & {
	embedded: {
		niceIndevFile?: IndevFile;
		niceIndevConvertedDocument?: IndevConvertedDocument;
	};
};

export type IndevConsultation = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	reference: string;
	title: string;
	consultationName: string;
	startDate: string;
	endDate: string;
	consultationType: string;
	resourceTitleId: string;
	projectType: string;
	technologyType: string;
	productTypeName: string;
	showExpressionOfInterestSubmissionQuestion: boolean;
	developedAs: string | null;
	relevantTo: string | null;
	consultationId: number;
	process: string;
	hasDocumentsWhichAllowConsultationComments: boolean;
	isCHTE: boolean;
	allowedRole: string;
	firstConvertedDocumentId: string | null;
	firstChapterSlugOfFirstConvertedDocument: string | null;
	partiallyUpdatedProjectReference: string | null;
	origProjectReference: string | null;
	areasOfInterestList: AreaOfInterest[];
	hidden: boolean;
};

export type IndevConsultationPanel = IndevPanel & {
	embedded: {
		niceIndevConsultation: IndevConsultation;
	};
};

export type IndevPanel = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevResourceList: {
			links: {
				self: [Link];
			};
			embedded: {
				niceIndevResource: IndevResource | IndevResource[];
			};
			eTag: ETag;
			hasResources: boolean;
		};
		niceIndevConsultation?: IndevConsultation;
	};
	eTag: ETag;
	title: string;
	panelType: string;
	legacyPanel: boolean;
	originalReference: string | null;
	updateReference: string | null;
	showPanel: boolean;
};

export type IndevTimeline = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
	additionalInfoLabel: string;
	additionalInfo: string;
	hidden: boolean;
};

export type IndevProjectTeam = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevLegacyStakeholder = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type IndevConsultee = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevCommentator = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	column1: string;
	column2: string;
};

export type IndevEmailEnquiry = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type IndevProcessHomepage = {
	links: {
		self: [Link];
	};
	etag: ETag;
	description: string;
	linkText: string;
} | null;

export type IndevProjectRelatedLink = {
	links: {
		self: [Link];
	};
	etag: ETag;
	column1: string;
	column2: string;
};

export type NiceIndevProjectRelatedLinkList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProjectRelatedLink:
			| IndevProjectRelatedLink
			| IndevProjectRelatedLink[];
	};
	etag: ETag;
};

export type IndevSchedule = {
	links: {
		self: [Link];
	};
	etag: ETag;
	column1: string;
	column2: string;
	additionalInfoLabel: string;
	additionalInfo: string;
	hidden: boolean;
};

export type NiceIndevProvisionalScheduleList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProvisionalSchedule: IndevSchedule | IndevSchedule[];
	};
	etag: ETag;
};

export type NiceIndevTimelineList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevTimeline: IndevTimeline | IndevTimeline[];
	};
};

export type NiceIndevProjectTeamList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevProjectTeam: IndevProjectTeam | IndevProjectTeam[];
	};
};

export type NiceIndevLegacyStakeholderList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevLegacyStakeholder:
			| IndevLegacyStakeholder
			| IndevLegacyStakeholder[];
	};
};

export type NiceIndevConsulteeList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevConsultee: IndevConsultee | IndevConsultee[];
	};
};

export type NiceIndevCommentatorList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevCommentator: IndevCommentator | IndevCommentator[];
	};
};

export type NiceIndevEmailEnquiryList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevEmailEnquiry: IndevEmailEnquiry | IndevEmailEnquiry[];
	};
};

export type IndevUpdate = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	updateReference: string;
};

export type NiceIndevFullUpdateList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevFullUpdate: IndevUpdate | IndevUpdate[];
	};
	eTag: ETag;
};

export type NiceIndevPartialUpdateList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevPartialUpdate: IndevUpdate | IndevUpdate[];
	};
	eTag: ETag;
};

export type IndevTopic = {
	links: {
		self: [Link];
	};
	eTag: ETag;
	item: string;
};

export type NiceIndevTopicList = {
	links: {
		self: [Link];
	};
	embedded: {
		niceIndevTopic: IndevTopic | IndevTopic[];
	};
	eTag: ETag;
};

export enum ProjectType {
	"NG" = "NG",
	"CSG" = "CSG",
	"CG" = "CG",
	"DG" = "DG",
	"MPG" = "MPG",
	"ES" = "ES",
	"HST" = "HST",
	"IPG" = "IPG",
	"KTT" = "KTT",
	"LGB" = "LGB",
	"MIB" = "MIB",
	"MTG" = "MTG",
	"PH" = "PH",
	"QS" = "QS",
	"TA" = "TA",
	"PSG" = "PSG",
	"SC" = "SC",
	"SG" = "SG",
	"PIP" = "PIP",
	"APG" = "APG",
	"APA" = "APA",
	"COV" = "COV",
	"COVM" = "COVM",
	"COVT" = "COVT",
	"COA" = "COA",
	"NGC" = "NGC",
	"IND" = "IND",
	"HTE" = "HTE",
	"HUB" = "HUB",
}

export type ProjectDetail = {
	links: {
		niceIndevStakeholderRegistration: Record<string, unknown>[];
		self: [Link];
	};
	embedded: {
		niceIndevFullUpdateList?: NiceIndevFullUpdateList;
		niceIndevPartialUpdateList?: NiceIndevPartialUpdateList;
		niceIndevTopicList?: NiceIndevTopicList;
		niceIndevPanelList: {
			links: {
				self: [Link];
			};
			embedded: {
				niceIndevPanel: IndevPanel | IndevPanel[];
			};
		};
		niceIndevProvisionalScheduleList?: NiceIndevProvisionalScheduleList;
		niceIndevTimelineList?: NiceIndevTimelineList;
		niceIndevProcessHomepage?: IndevProcessHomepage;
		niceIndevProjectTeamList?: NiceIndevProjectTeamList;
		niceIndevProjectRelatedLinkList?: NiceIndevProjectRelatedLinkList;
		niceIndevLegacyStakeholderList?: NiceIndevLegacyStakeholderList;
		niceIndevConsulteeList?: NiceIndevConsulteeList;
		niceIndevCommentatorList?: NiceIndevCommentatorList;
		niceIndevEmailEnquiryList?: NiceIndevEmailEnquiryList;
	};
	eTag: ETag;
	summary: string | null;
	description: string | null;
	referralDate: string | null;
	suspendDiscontinuedReason: string | null;
	suspendDiscontinuedUrl: string | null;
	suspendDiscontinuedUrlText: string | null;
	legacyModel: boolean;
	productReference: string | null;
	evidenceAssessmentGroup: string | null;
	reference: string;
	projectType: ProjectType | null;
	technologyType: string | null;
	productTypeName: string | null;
	process: string;
	alert: string | null;
	title: string;
	content: string | null;
	status: ProjectStatus;
	projectGroup: ProjectGroup;
	publishedDate: string;
	lastModifiedDate: string;
	firstGoLiveDate: string;
	createdDate: string;
	developedAs: string;
	relevantTo: string | null;
	idNumber: string | null;
	areasOfInterestList: [];
	topicSelectionDecision: string;
	topicSelectionReason: string | null;
	topicSelectionDecisionDate: string | null;
	topicSelectionFurtherInfo: string | null;
	indicatorSubTypes: [];
};

export type resourceInPageNavLink = {
	slug: string;
	title: string;
};

export type niceIndevConvertedDocumentChapter = {
	slug: string;
	title: string;
	href: string;
};

export type niceIndevConvertedDocumentSection = {
	slug: string;
	title: string;
};

export type niceIndevConvertedDocument = {
	_links?: {
		self: [
			{
				href: string;
			}
		];
	};
	eTag?: string | null;
	content: string;
	pdfLink: string | null;
	title?: string;
	chapterTitle?: string;
	chapters: niceIndevConvertedDocumentChapter[];
	sections?: niceIndevConvertedDocumentSection[];
};
