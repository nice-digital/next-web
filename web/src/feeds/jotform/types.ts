export type FormID = `${number}`;

export type Duration = `${number}ms`;

export type NotFoundResponse = {
	responseCode: 404;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type NotAuthorizedResponse = {
	responseCode: 401;
	message: string;
	content: "";
	duration: Duration;
	info: string;
};

export type FormStatus = "ENABLED" | "DISABLED";

export type FormType = "LEGACY" | "CARD";

export type FormSuccessResponse = {
	responseCode: 200;
	message: "success";
	content: {
		id: string;
		username: string;
		title: string;
		height: number;
		status: FormStatus;
		/** Format like "2023-03-21 09:27:18" */
		created_at: string;
		updated_at: string;
		last_submission: string;
		new: `${number}`;
		count: `${number}`;
		type: FormType;
		favorite: `${number}`;
		archived: `${number}`;
		url: string;
	};
	duration: Duration;
};

export type GetFormResponse =
	| NotAuthorizedResponse
	| NotFoundResponse
	| FormSuccessResponse;
