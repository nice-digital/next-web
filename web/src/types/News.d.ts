import {
	type PodcastStoryblok,
	type BlogPostStoryblok,
	type NewsArticleStoryblok,
	type InDepthArticleStoryblok,
} from "./storyblok";

export type NewsStory =
	| BlogPostStoryblok
	| NewsArticleStoryblok
	| PodcastStoryblok
	| InDepthArticleStoryblok;
