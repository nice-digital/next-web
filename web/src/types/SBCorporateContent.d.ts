import { ExtendedSBLink } from "@/components/Storyblok/StoryblokSectionNav/utils/Utils";

export type SlugCatchAllSuccessProps = {
	story: ISbStoryData<InfoPageStoryblok | CategoryNavigationStoryblok>;
	breadcrumbs: Breadcrumb[];
	component: string;
	tree: ExtendedSBLink[];
	slug: string;
};

export type SlugCatchAllErrorProps = {
	error: string;
};

export type SlugCatchAllProps =
	| SlugCatchAllSuccessProps
	| SlugCatchAllErrorProps;
