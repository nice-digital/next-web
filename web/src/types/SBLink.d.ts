// Responses from the links API, which Storyblok don't supply types for (yet)
// https://www.storyblok.com/docs/api/content-delivery/v2#core-resources/links/links

export type SBLink = {
	id: number;
	slug: string;
	name: string;
	is_folder: boolean;
	parent_id: number | null;
	published: boolean;
	position: number;
	uuid: string;
	is_startpage: boolean;
	real_path: string;
	path: string | null;
};

// The API response consists of an arbitrary number of properties
// The key of each property is a guid; the value is the link object itself
export type SBLinkRecord = Record<string, SBLink>;

// This is what we actually get back from the API
export type SBLinkResponse = {
	[SBLinkRecord];
};
