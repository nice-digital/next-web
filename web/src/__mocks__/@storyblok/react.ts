export const storyblokInit = (): null => null;

export const apiPlugin = (): null => null;
