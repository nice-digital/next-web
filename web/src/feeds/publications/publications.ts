import { serverRuntimeConfig } from "@/config";

import { getFeedBodyCached, getFeedBodyUnCached, getResponseStream } from "../";

import {
	AreaOfInterest,
	AreaOfInterestTypes,
	ErrorResponse,
	FeedPath,
	ChapterHTMLContent,
	IndicatorSubType,
	IndicatorSubTypes,
	ProductDetail,
	ProductListLite,
	ProductLite,
	ProductType,
	ProductTypes,
	ResourceDetail,
	RelatedResourceList,
	IndicatorMappings,
	IndicatorMapping,
} from "./types";

export * from "./types";

const cacheKeyPrefix = "publications",
	{ defaultTTL, longTTL } = serverRuntimeConfig.cache,
	{ origin, apiKey } = serverRuntimeConfig.feeds.publications;

/**
 * Gets a list of products from the 'products lite' endpoint.
 */
export const getAllProducts = async (): Promise<ProductLite[]> =>
	await getFeedBodyCached<ProductLite[]>(
		cacheKeyPrefix,
		FeedPath.ProductsLite,
		defaultTTL,
		async () =>
			(
				await getFeedBodyUnCached<ProductListLite>(
					origin,
					FeedPath.ProductsLite,
					apiKey
				)
			).embedded.productListLite.embedded.productLite.map((product) => {
				// Discard unneeded properties on products to make what we're storing in cache a lot smaller.
				// This means we're essentially storing a ProductLite in cache rather than a ProductLiteRaw.
				// In perf tests this saved ~30% off the cache load time from the file system (once you factor in deserialization, file access times etc).
				delete (product as Partial<typeof product>).eTag;
				delete (product as Partial<typeof product>).links;
				return product;
			})
	);

/**
 * Gets _all_ product types.
 *
 * Note: there's no pre-filter so it includes both enabled _and_ disabled product types.
 */
export const getAllProductTypes = async (): Promise<ProductType[]> =>
	await getFeedBodyCached<ProductType[]>(
		cacheKeyPrefix,
		FeedPath.ProductTypes,
		longTTL,
		async () =>
			(
				await getFeedBodyUnCached<ProductTypes>(
					origin,
					FeedPath.ProductTypes,
					apiKey
				)
			).productTypes
	);

/**
 * Gets _all_ areas of interest e.g. Covid-19 and Antimicrbial prescribing.
 *
 * Note: there's no pre-filter so it includes both enabled _and_ disabled areas of interest.
 */
export const getAllAreasOfInterest = async (): Promise<AreaOfInterest[]> =>
	await getFeedBodyCached<AreaOfInterest[]>(
		cacheKeyPrefix,
		FeedPath.AreasOfInterest,
		longTTL,
		async () =>
			(
				await getFeedBodyUnCached<AreaOfInterestTypes>(
					origin,
					FeedPath.AreasOfInterest,
					apiKey
				)
			).areaOfInterestTypes
	);

/**
 * Gets _all_ indicator sub types e.g. Clinical commissioning group indicator.
 *
 * Note: there's no pre-filter so it includes both enabled _and_ disabled indicator sub types.
 */
export const getAllIndicatorSubTypes = async (): Promise<IndicatorSubType[]> =>
	await getFeedBodyCached<IndicatorSubType[]>(
		cacheKeyPrefix,
		FeedPath.IndicatorSubTypes,
		longTTL,
		async () =>
			(
				await getFeedBodyUnCached<IndicatorSubTypes>(
					origin,
					FeedPath.IndicatorSubTypes,
					apiKey
				)
			).indicatorSubTypes
	);

/**
 * Gets the indicator sub type object from the given identifier prefix
 *
 * @param identifierPrefix The identifier prefix of the indicator sub type to find e.g. CCG, GPIQ, GPINQ, NLQ etc
 * @returns The indicator sub type object if found, otherwise null
 */
export const getIndicatorSubType = async (
	identifierPrefix: string
): Promise<IndicatorSubType | null> =>
	(await getAllIndicatorSubTypes()).find(
		(subType) => subType.identifierPrefix === identifierPrefix
	) || null;

/**
 * Gets _all_ indicator mappings (redirects).
 */
export const getIndicatorMappings = async (): Promise<IndicatorMapping[]> =>
	await getFeedBodyCached<IndicatorMapping[]>(
		cacheKeyPrefix,
		FeedPath.IndicatorMappings,
		defaultTTL,
		async () =>
			(
				await getFeedBodyUnCached<IndicatorMappings>(
					origin,
					FeedPath.IndicatorMappings,
					apiKey
				)
			).mappings
	);

/**
 * Gets a full product detail response from publications from the given product id.
 *
 * @returns The full product details, or `null` if the product can't be found
 */
export const getProductDetail = async (
	productId: string
): Promise<ProductDetail | null> =>
	await getFeedBodyCached<ProductDetail | null>(
		cacheKeyPrefix,
		FeedPath.ProductDetail + productId,
		longTTL,
		async () => {
			const response = await getFeedBodyUnCached<ProductDetail | ErrorResponse>(
				origin,
				FeedPath.ProductDetail + productId,
				apiKey
			);

			return isSuccessResponse(response) ? response : null;
		}
	);

/**
 * Gets chapter HTML.
 *
 */
export const getChapterContent = async (
	chapterHref: string
): Promise<ChapterHTMLContent | null> => {
	return getFeedBodyCached<ChapterHTMLContent | null>(
		cacheKeyPrefix,
		chapterHref,
		longTTL,
		async () => {
			const response = await getFeedBodyUnCached<
				ChapterHTMLContent | ErrorResponse
			>(origin, chapterHref, apiKey);

			return isSuccessResponse(response) ? response : null;
		}
	);
};

/**
 * Fetches a full resource response from publications.
 *
 * Returns null if the resource can't be found.
 *
 * @param resource The related resource, or null if the resource can't be found
 */
export const getResourceDetail = async (
	resource: RelatedResourceList
): Promise<ResourceDetail | null> => {
	const { url } = resource;

	return getFeedBodyCached<ResourceDetail | null>(
		cacheKeyPrefix,
		url,
		longTTL,
		async () => {
			const response = await getFeedBodyUnCached<
				ResourceDetail | ErrorResponse
			>(origin, url, apiKey);

			return isSuccessResponse(response) ? response : null;
		}
	);
};

/**
 * Fetches full resources responses from publications for each of the given resources.
 *
 * It will throw an error if any of the resources can't be found
 */
export const getResourceDetails = async (
	resources: RelatedResourceList[]
): Promise<ResourceDetail[]> => {
	const fullResources = await Promise.all(
		resources.map(
			(relatedResource) =>
				new Promise<{
					relatedResource: RelatedResourceList;
					resourceDetail: ResourceDetail | null;
				}>((resolve) => {
					getResourceDetail(relatedResource).then((resourceDetail) => {
						resolve({
							relatedResource,
							resourceDetail,
						});
					});
				})
		)
	);

	// We can't 100% guarantee all related resources always still exist so handle if some can't be found
	const failedResources = fullResources
		.filter((r) => r.resourceDetail === null)
		.map((r) => r.relatedResource);

	if (failedResources.length)
		throw Error(
			`Could not load resources ${failedResources.map((r) => r.uid)}`
		);

	return fullResources
		.map((r) => r.resourceDetail)
		.filter((r): r is ResourceDetail => r !== null);
};

export const getImage = async (url: string): Promise<Buffer> =>
	await getFeedBodyCached<Buffer>(cacheKeyPrefix, url, longTTL, async () => {
		// Construct the full URL and headers for the request
		const logourl = origin + url;

		const response = await fetch(logourl, {
			method: "GET",
			headers: {
				"Api-Key": apiKey,
				Accept: "application/json",
			},
		});

		// Check if the response is OK
		if (!response.ok) {
			throw new Error(
				`Failed to fetch image from ${url}: ${response.statusText}`
			);
		}

		// Convert the response stream to a Blob
		const blob = await response.blob();

		let buffer = Buffer.from(await blob.arrayBuffer());

		return buffer;
	});

/**
 * Gets a stream of a file from publications.
 *
 * @param filePath The relative path of the endpoint that serves file content, e.g. `/feeds/downloads/737585a0-dad7-4a37-875b-b30b09c3fdc3`
 * @returns A readable stream of the file contents
 */
export const getFileStream = async (
	filePath: string
): Promise<ReturnType<typeof getResponseStream>> =>
	getResponseStream(origin, filePath, apiKey);

/**
 * A user-defined type guard for checking whether a given response is an error or not
 *
 * @param response The response
 * @returns True if this is an error response otherwise false
 */
export function isErrorResponse<TValidResponse>(
	response: TValidResponse | ErrorResponse
): response is ErrorResponse {
	return (response as ErrorResponse).statusCode !== undefined;
}

/**
 * A user-defined type guard for checking whether a given response is not an error.
 * The opposite of `isErrorResponse`
 *
 * @param response The response
 * @returns True if this is an error response otherwise false
 */
export function isSuccessResponse<TValidResponse>(
	response: TValidResponse | ErrorResponse
): response is TValidResponse {
	return (response as ErrorResponse).statusCode === undefined;
}
