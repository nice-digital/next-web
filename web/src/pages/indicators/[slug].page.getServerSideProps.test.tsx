import { render } from "@testing-library/react";
import axios from "axios";

import { ProductGroup, ProductTypeAcronym } from "@/feeds/publications/types";

import IndicatorsDetailsPage, {
	getServerSideProps,
	IndicatorsDetailsPageProps,
} from "./[slug].page";

import type { GetServerSidePropsContext } from "next";

// jest.mock("axios");
// const mockedAxios = axios as jest.Mocked<typeof axios>;

jest.mock("@/feeds/publications/publications", () => {
	const originalModule = jest.requireActual(
		"@/feeds/publications/publications"
	);

	//Mock the getProductDetail export
	return {
		__esModule: true,
		...originalModule,
		getProductDetail: jest.fn(() => ({
			title: "Test title 1",
			id: "IND1",
		})),
	};
});

const props: IndicatorsDetailsPageProps = {
	slug: "/ind-1-test-title-1",
	id: "IND1",
	product: { title: "/test", id: "IND1", productType: "IND" },
	productType: {
		_links: {
			self: [{}],
		},
		eTag: null,
		lastModified: "2022-07-07T00:00:00",
		enabled: true,
		name: "NICE indicator",
		pluralName: "NICE indicators",
		identifierPrefix: ProductTypeAcronym.IND,
		group: ProductGroup.Other,
		parent: "",
	},
} as unknown as IndicatorsDetailsPageProps;

describe("IndicatorDetailPage", () => {
	it("should match snapshot for main content", () => {
		render(<IndicatorsDetailsPage {...props} />);

		expect(document.body).toMatchSnapshot();
	});
});

describe("getGetServerSidePropsFunc", () => {
	it("should return a correct props when supplied with an id", async () => {
		const result = await getServerSideProps({
			params: { slug: "ind1-test-title-1" },
		} as unknown as GetServerSidePropsContext);

		expect(result).toStrictEqual({
			props: {
				slug: "ind1-test-title-1",
				id: "ind1",
				product: { id: "IND1", title: "Test title 1" },
			},
		});
	});

	describe("Redirects", () => {
		it("should return permanent redirect object URL with incorrect title", async () => {
			const redirectResult = await getServerSideProps({
				params: { slug: "ind1-incorrect-slug-title" },
			} as unknown as GetServerSidePropsContext);

			expect(redirectResult).toStrictEqual({
				redirect: {
					destination: "/indicators/ind1-test-title-1",
					permanent: true,
				},
			});
		});
		it("should return notFound if Id doesn't exist", async () => {
			const redirectResult = await getServerSideProps({
				params: { slug: "nonExistingId99-test-title-1" },
			} as unknown as GetServerSidePropsContext);

			expect(redirectResult).toStrictEqual({ notFound: true });
		});

		it("should return notFound if an unhyphenated slug is used", async () => {
			const redirectResult = await getServerSideProps({
				params: { slug: "slugwithouthyphenation" },
			} as unknown as GetServerSidePropsContext);

			expect(redirectResult).toStrictEqual({ notFound: true });
		});

		//TODO check logic for any missing cases
	});
});
