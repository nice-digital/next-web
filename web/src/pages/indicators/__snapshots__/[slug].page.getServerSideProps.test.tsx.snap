// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndicatorDetailPage should match snapshot for main content 1`] = `
<body>
  <div>
    <nav
      aria-label="Breadcrumbs"
      role="navigation"
    >
      <ol
        class="breadcrumbs"
      >
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/"
          >
            Home
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/standards-and-indicators"
          >
            Standards and Indicators
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <a
            href="/standards-and-indicators/indicators"
          >
            Indicators
          </a>
        </li>
        <li
          class="breadcrumbs__crumb"
        >
          <span>
            IND1
          </span>
        </li>
      </ol>
    </nav>
    <script
      type="application/ld+json"
    >
      {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Home"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/standards-and-indicators",
        "name": "Standards and Indicators"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/standards-and-indicators/indicators",
        "name": "Indicators"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "IND1"
      }
    }
  ]
}
    </script>
    <div
      class="page-header"
      id="content-start"
    >
      <h1
        class="page-header__heading page-header__heading--alt"
      >
        test title
      </h1>
      <p
        class="page-header__lead"
      >
        <div
          class="leadMeta"
        >
          <span>
            NICE indicator
          </span>
          <span>
            IND1
          </span>
          <span>
            Published:
            <time
              datetime="2022-09-08"
            >
              8 September 2022
            </time>
          </span>
        </div>
      </p>
    </div>
    <div
      class="grid grid--loose"
    >
      <section
        aria-label="Chapters"
        data-g="12 md:4 lg:3"
      />
      <section
        data-g="12 md:8 lg:9"
      />
    </div>
  </div>
</body>
`;
